### PR TITLE
Plumbing refactor to allow the usage of Dart_CreateIsolateInGroup

### PR DESCRIPTION
**Pull request by gaaclarke**
## Description

This should be a noop refactor that plumbs spawning all the way down to the dart isolate so that it can use Dart_CreateIsolateInGroup if necessary.

## Related Issues

https://github.com/flutter/flutter/issues/72025

## Tests

I added the following tests:

This is just a refactor, no new or changed functionality, existing tests should cover it.

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [contributor guide] and followed the process outlined there for submitting PRs.
- [x] I signed the [CLA].
- [x] I read and followed the [C++, Objective-C, Java style guides] for the engine.
- [x] I read the [tree hygiene] wiki page, which explains my responsibilities.
- [x] I updated/added relevant documentation.
- [x] All existing and new tests are passing.
- [x] I am willing to follow-up on review comments in a timely manner.

## Reviewer Checklist

- [x] I have submitted any presubmit flakes in this PR using the [engine presubmit flakes form] before re-triggering the failure.

## Breaking Change

Did any tests fail when you ran them? Please read [handling breaking changes].

- [x] No, no existing tests failed, so this is *not* a breaking change.
- [ ] Yes, this is a breaking change. *If not, delete the remainder of this section.*
   - [ ] I wrote a design doc: https://flutter.dev/go/template *Replace this with a link to your design doc's short link*
   - [ ] I got input from the developer relations team, specifically from: *Replace with the names of who gave advice*
   - [ ] I wrote a migration guide: https://flutter.dev/go/breaking-changes-template *Replace this with a link to a pull request that adds the migration guide to https://flutter.dev/docs/release/breaking-changes*

[issue database]: https://github.com/flutter/flutter/issues
[contributor guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[testing the engine]: https://github.com/flutter/flutter/wiki/Testing-the-engine
[C++, Objective-C, Java style guides]: https://github.com/flutter/engine/blob/master/CONTRIBUTING.md#style
[CLA]: https://cla.developers.google.com/
[tree hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[handling breaking changes]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[engine presubmit flakes form]: https://forms.gle/Wc1VyFRYJjQTH6w5A

### CODE REVIEW

**Review by chinmaygarde — COMMENTED** [review]
This breaks the existing behavior that terminating the root isolate also terminates any isolates spawned by it. It should be possible to create distinct isolate groups with shared resources. Also, what resources are being duplicated during isolate creation?

File: runtime/dart_isolate.h [hunk 1]
```diff
@@ -232,7 +231,29 @@ class DartIsolate : public UIDartState {
...
-      std::shared_ptr<VolatilePathTracker> volatile_path_tracker);
+      std::shared_ptr<VolatilePathTracker> volatile_path_tracker,
+      const DartIsolate* spawning_isolate = nullptr);
+
+  //----------------------------------------------------------------------------
+  /// @brief     Creates a running DartIsolate who shares as many resources as
+  ///            possible with the caller DartIsolate.  This allows them to
+  ///            occupy less memory together.
+  /// @attention Only certain setups can take advantage of the most savings
+  ///            currently, release builds currently.
```
Comment (xster) [inline runtime/dart_isolate.h:242]: @mkustermann where do you estimate you guys will land by the next Flutter stable?

File: runtime/dart_isolate.h [hunk 2]
```diff
@@ -232,7 +231,29 @@ class DartIsolate : public UIDartState {
...
       std::optional<std::string> dart_entrypoint_library,
       std::unique_ptr<IsolateConfiguration> isolate_configration,
-      std::shared_ptr<VolatilePathTracker> volatile_path_tracker);
+      std::shared_ptr<VolatilePathTracker> volatile_path_tracker,
+      const DartIsolate* spawning_isolate = nullptr);
+
+  //----------------------------------------------------------------------------
+  /// @brief     Creates a running DartIsolate who shares as many resources as
+  ///            possible with the caller DartIsolate.  This allows them to
+  ///            occupy less memory together.
```
Comment (xster) [inline runtime/dart_isolate.h:240]: Describe their and their components' lifecycles. Describe restrictions (i.e. they must be from the same aot).

File: shell/common/engine.h [hunk 3]
```diff
@@ -354,6 +354,19 @@ class Engine final : public RuntimeDelegate,
          fml::WeakPtr<SnapshotDelegate> snapshot_delegate,
          std::shared_ptr<VolatilePathTracker> volatile_path_tracker);
 
+  //----------------------------------------------------------------------------
+  /// @brief      Create a Engine that shares as many resources as
+  ///             possible with the calling Engine such that together
+  ///             they occupy less memory.
```
Comment (xster) [inline shell/common/engine.h:360]: Same comment for doc. Can use more details.

File: shell/common/shell.h [hunk 4]
```diff
@@ -98,6 +98,20 @@ class Shell final : public PlatformView::Delegate,
  public:
   template <class T>
   using CreateCallback = std::function<std::unique_ptr<T>(Shell&)>;
+  typedef std::function<std::unique_ptr<Engine>(
```
Comment (xster) [inline shell/common/shell.h:101]: Kinda does the same thing as the above. Can we make the names sound similar like EngineCreateCallback or something?

File: shell/common/shell.h [hunk 5]
```diff
@@ -169,6 +183,27 @@ class Shell final : public PlatformView::Delegate,
...
+  ///
+  static std::unique_ptr<Shell> Create(
+      TaskRunners task_runners,
+      const PlatformData platform_data,
+      Settings settings,
+      fml::RefPtr<const DartSnapshot> isolate_snapshot,
+      const CreateCallback<PlatformView>& on_create_platform_view,
+      const CreateCallback<Rasterizer>& on_create_rasterizer,
+      DartVMRef vm,
+      const EngineMaker& engine_maker);
```
Comment (xster) [inline shell/common/shell.h:205]: Same here. Could this be something like on_create_engine for consistency?

**Review by xster — COMMENTED** [review]
LGTM. Please wait for @chinmaygarde to review.

**Comment by gaaclarke** [comment]
> what resources are being duplicated during isolate creation?

The isolate group, so the heap and GC and the loaded executable.

This PR isn't making any functional changes.  The top level APIs behaves the same.  A spawned flutter engine and its isolate should live beyond its spawning engine, this matches the semantics of isolate groups and matches how engine groups will work.

I've added a comment to where the functional change will be made, conditionally we will be calling Dart_CreateIsolateInGroup instead of Dart_CreateIsolateGroup.  I avoided turning it fully on yet until I can finish all the phase 1 work.  I brought this over because this refactor makes other phase 1 work possible.  You can see it in action in the prototype https://github.com/gaaclarke/engine/tree/spawn-engine

File: runtime/dart_isolate.h (hunk 2 above)
Comment (gaaclarke) [reply runtime/dart_isolate.h]: Done

File: shell/common/engine.h (hunk 3 above)
Comment (gaaclarke) [reply shell/common/engine.h]: Done, mostly linked to other documentation.

File: shell/common/shell.h (hunk 4 above)
Comment (gaaclarke) [reply shell/common/shell.h:101]: done

File: shell/common/shell.h (hunk 5 above)
Comment (gaaclarke) [reply shell/common/shell.h]: good idea, done

**Comment by gaaclarke** [comment]
@chinmaygarde This is a whole new API so there isn't any expectations.  The behavior you are describing is definitive to the work we are doing.

The heap isn't per isolate, it is per group although there are safeguards that disallow them from directly accessing memory allocated to a different isolate (there is plans to change that).  With respect to the loaded program it isn't a problem with occupied memory but the time to create a new isolate.  Loading the executable takes a sizable chunk of the time to create a new isolate, in this case it is just wasted effort.

All this work is outlined in this issue: https://github.com/dart-lang/sdk/issues/36097  I think there is a design doc floating around, too.

File: shell/common/shell.h [hunk 6]
```diff
@@ -169,6 +183,27 @@ class Shell final : public PlatformView::Delegate,
       CreateCallback<PlatformView> on_create_platform_view,
       CreateCallback<Rasterizer> on_create_rasterizer);
 
+  //----------------------------------------------------------------------------
+  /// @brief      Creates a shell instance using the provided settings.
+  /// @details    This version of Create can take in a running DartVMRef and a
+  ///             function that defines how the Shell's Engine should be
+  ///             created.
+  /// @param[in]  vm             A running DartVMRef where this Shell's Dart
```
Comment (xster) [inline shell/common/shell.h:191]: Do you want to briefly describe how one typically gets a DartVMRef?

**Review by xster — APPROVED** [review]

File: runtime/dart_isolate.h [hunk 7]
```diff
@@ -232,7 +231,33 @@ class DartIsolate : public UIDartState {
...
+      const DartIsolate* spawning_isolate = nullptr);
+
+  //----------------------------------------------------------------------------
+  /// @brief     Creates a running DartIsolate who shares as many resources as
+  ///            possible with the caller DartIsolate.  This allows them to
+  ///            occupy less memory together and to be created faster.
+  /// @details   Shared components will be destroyed when the last DartIsolate
+  ///            is destroyed.  SpawnIsolate can only be used to create
+  ///            DartIsolates whose executable code is shared with the calling
+  ///            DartIsolate.
```
[the remark this answers is not in the document]
Comment (gaaclarke) [reply runtime/dart_isolate.h:244]: I'm not sure of the hard requirements of Dart_CreateIsolateInGroup.  It isn't spelled out in its documentation.

File: shell/common/shell.h (hunk 6 above)
Comment (gaaclarke) [reply shell/common/shell.h:191]: I don't know if we should point to DartVMRef::Create, it's pretty straightforward

File: runtime/dart_isolate.h [hunk 8]
```diff
@@ -232,7 +231,33 @@ class DartIsolate : public UIDartState {
       std::optional<std::string> dart_entrypoint,
       std::optional<std::string> dart_entrypoint_library,
       std::unique_ptr<IsolateConfiguration> isolate_configration,
-      std::shared_ptr<VolatilePathTracker> volatile_path_tracker);
+      std::shared_ptr<VolatilePathTracker> volatile_path_tracker,
+      const DartIsolate* spawning_isolate = nullptr);
```
[the remark this answers is not in the document]
Comment (gaaclarke) [reply runtime/dart_isolate.h:237]: Done

File: runtime/dart_isolate.h [hunk 9]
```diff
@@ -232,7 +231,33 @@ class DartIsolate : public UIDartState {
...
+
+  //----------------------------------------------------------------------------
+  /// @brief     Creates a running DartIsolate who shares as many resources as
+  ///            possible with the caller DartIsolate.  This allows them to
+  ///            occupy less memory together and to be created faster.
+  /// @details   Shared components will be destroyed when the last DartIsolate
```
[the remark this answers is not in the document]
Comment (gaaclarke) [reply runtime/dart_isolate.h]: updated

File: runtime/dart_isolate.h [hunk 10]
```diff
@@ -232,7 +231,33 @@ class DartIsolate : public UIDartState {
...
+  ///            is destroyed.  SpawnIsolate can only be used to create
+  ///            DartIsolates whose executable code is shared with the calling
+  ///            DartIsolate.
+  /// @attention Only certain setups can take advantage of the most savings
+  ///            currently, AOT specifically.
+  /// @return    A running DartIsolate.
```
[the remark this answers is not in the document]
Comment (gaaclarke) [reply runtime/dart_isolate.h]: done

File: runtime/dart_isolate.h [hunk 11]
```diff
@@ -232,7 +231,33 @@ class DartIsolate : public UIDartState {
...
+  //----------------------------------------------------------------------------
+  /// @brief     Creates a running DartIsolate who shares as many resources as
+  ///            possible with the caller DartIsolate.  This allows them to
+  ///            occupy less memory together and to be created faster.
+  /// @details   Shared components will be destroyed when the last DartIsolate
+  ///            is destroyed.  SpawnIsolate can only be used to create
+  ///            DartIsolates whose executable code is shared with the calling
+  ///            DartIsolate.
+  /// @attention Only certain setups can take advantage of the most savings
+  ///            currently, AOT specifically.
```
[the remark this answers is not in the document]
Comment (gaaclarke) [reply runtime/dart_isolate.h:248]: done

File: shell/common/engine.h [hunk 12]
```diff
@@ -354,6 +354,22 @@ class Engine final : public RuntimeDelegate,
          fml::WeakPtr<SnapshotDelegate> snapshot_delegate,
          std::shared_ptr<VolatilePathTracker> volatile_path_tracker);
 
+  //----------------------------------------------------------------------------
+  /// @brief      Create a Engine that shares as many resources as
+  ///             possible with the calling Engine such that together
+  ///             they occupy less memory and be created faster.
+  /// @details    This method ultimately calls DartIsolate::SpawnIsolate to make
+  ///             sure resources are shared.
```
[the remark this answers is not in the document]
Comment (gaaclarke) [reply shell/common/engine.h]: Done.

File: runtime/dart_isolate.h [hunk 13]
```diff
@@ -232,7 +231,33 @@ class DartIsolate : public UIDartState {
       std::optional<std::string> dart_entrypoint,
       std::optional<std::string> dart_entrypoint_library,
       std::unique_ptr<IsolateConfiguration> isolate_configration,
-      std::shared_ptr<VolatilePathTracker> volatile_path_tracker);
+      std::shared_ptr<VolatilePathTracker> volatile_path_tracker,
+      const DartIsolate* spawning_isolate = nullptr);
+
+  //----------------------------------------------------------------------------
+  /// @brief     Creates a running DartIsolate who shares as many resources as
```
[the remark this answers is not in the document]
Comment (gaaclarke) [reply runtime/dart_isolate.h:240]: I don't think that is much of an improvement.  Hopefully the other edits cover the information you are trying to convey with that.

**Review by chinmaygarde — COMMENTED** [review]
Mostly looks good. Please add tests for `Engine::Spawn` to `shell_unittests` and for `DartIsolate::SpawnIsolate` to `runtime_unittests`.

**Comment by gaaclarke** [comment]
> Mostly looks good. Please add tests for `Engine::Spawn` to `shell_unittests` and for `DartIsolate::SpawnIsolate` to `runtime_unittests`.

A dependent PR has a unit test for Engine::Spawn https://github.com/flutter/engine/pull/23603   SpawnIsolate has test coverage via the golden image test for spawning engines and the FlutterEngineTests of objective-c.  I'll look into seeing if I can get a unit test directly on SpawnIsolate.

**Comment by chinmaygarde** [comment]
> I'll look into seeing if I can get a unit test directly on SpawnIsolate.

You should be able to base the test on [`DartIsolateTest::RootIoslateCreationAndShutdown`](https://github.com/flutter/engine/blob/d2b815479045b9e7f1c17e8eaeae6a414336a15d/runtime/dart_isolate_unittests.cc#L37). Take that isolate and spawn another one using that as the base.

SGTM about the test in the dependent PR.

**Pull request merged**